aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,404 @@
+            })
+        }
+    }
+
+    /// Return the algorithm associated with the given KEM private key.
+    #[must_use]
+    pub fn algorithm(&self) -> &'static KemAlgorithm {
+        self.algorithm
+    }
+
+    /// Computes the KEM public key from the KEM private key
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn compute_public_key(&self) -> Result<KemPublicKey, Unspecified> {
+        let mut pubkey_bytes = vec![0u8; self.algorithm.public_key_size()];
+        unsafe {
+            if 1 != EVP_PKEY_get_raw_public_key(
+                *self.pkey,
+                pubkey_bytes.as_mut_ptr(),
+                &mut self.algorithm.public_key_size(),
+            ) {
+                return Err(Unspecified);
+            }
+
+            let pubkey_ctx_copy = LcPtr::new(EVP_PKEY_kem_new_raw_public_key(
+                self.algorithm.id.nid(),
+                pubkey_bytes.as_mut_ptr(),
+                self.algorithm.public_key_size(),
+            ))?;
+
+            Ok(KemPublicKey {
+                algorithm: self.algorithm,
+                pkey: pubkey_ctx_copy,
+                pub_key: pubkey_bytes.into(),
+            })
+        }
+    }
+
+    /// Performs the decapsulate operation using the current KEM private key on the given ciphertext.
+    ///
+    /// `ciphertext` is the ciphertext generated by the encapsulate operation using the KEM public key
+    /// associated with the current KEM private key.
+    ///
+    /// After the decapsulation is finished, `decapsulate` calls `kdf` with the raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn decapsulate<F, R>(&self, ciphertext: &mut [u8], kdf: F) -> Result<R, Unspecified>
+    where
+        F: FnOnce(&[u8]) -> Result<R, Unspecified>,
+    {
+        unsafe {
+            let ctx = LcPtr::new(EVP_PKEY_CTX_new(*self.pkey, null_mut()))?;
+            let mut shared_secret: Vec<u8> = vec![0u8; self.algorithm.shared_secret_size()];
+
+            if 1 != EVP_PKEY_decapsulate(
+                *ctx,
+                shared_secret.as_mut_ptr(),
+                &mut self.algorithm.secret_key_size(),
+                ciphertext.as_mut_ptr(),
+                ciphertext.len(),
+            ) {
+                return Err(Unspecified);
+            }
+            kdf(&shared_secret)
+        }
+    }
+
+    /// Creates a new KEM private key from raw bytes
+    ///
+    /// `alg` is the `KemAlgorithm` to be associated with the generated `KemPrivateKey`
+    ///
+    /// `bytes` is a slice of raw bytes representing a `KemPrivateKey`
+    ///
+    /// # Errors
+    /// `error::KeyRejected` when operation fails during key creation.
+    ///
+    pub fn new(alg: &'static KemAlgorithm, bytes: &[u8]) -> Result<Self, KeyRejected> {
```

Comment:
This comment I had seems to have gotten lost: 
https://github.com/aws/aws-lc-rs/pull/174#discussion_r1248317298